addons/HotelPvP/scripts/Game/AFM_HotelPvP/GameMode/AFM_FactoryPvPGameMode.c

```diff
@@ -0,0 +1,442 @@
+				{
+					GameEndDefendersWin();
+				}
+				else if (killerFactionKey == m_sDefenderFactionKey)
+				{
+					GameEndAttackersWin();
+				}
+				else
+				{
+					PrintFormat("Unsuported killer faction! %1", killerFactionKey, level:LogLevel.ERROR);
+					GameEndTie();
+				}
+			}
+		}
+		
+		GameEndCheck();
+		Replication.BumpMe();
+	}
+	
+	override void EOnInit(IEntity owner)
+	{
+		super.EOnInit(owner);
+		
+		if (SCR_Global.IsEditMode())
+			return;
+		
+		m_FactionManager = SCR_FactionManager.Cast(GetGame().GetFactionManager());
+		if (!m_FactionManager)
+		{
+			Print("Faction manager component is missing!", LogLevel.ERROR);
+		}
+		
+		ChimeraWorld world = GetGame().GetWorld();
+		m_fStartTimestamp = world.GetServerTimestamp().PlusSeconds(m_iGameStartDelaySeconds);
+		m_fVictoryTimestamp = m_fStartTimestamp.PlusSeconds(m_iGameOverTimeMinutes * 60);
+		GetGame().GetCallqueue().CallLater(StartGame, m_iGameStartDelaySeconds * 1000);
+		GetGame().GetCallqueue().CallLater(TimeoutGameEnd, m_iGameOverTimeMinutes * 60 * 1000 + m_iGameStartDelaySeconds * 1000, false);
+	}
+	
+	void StartGame()
+	{
+		if (!IsMaster())
+			return;
+		
+		Print("Gamemode starting");
+		DisableRestrictionZones();
+		GetGame().GetCallqueue().CallLater(GameEndCheck, 1000 * 5, true);
+		Rpc(RPC_DoStartGame);
+		RPC_DoStartGame();
+		Replication.BumpMe();
+	}
+	
+	[RplRpc(RplChannel.Reliable, RplRcver.Broadcast)]
+	void RPC_DoStartGame()
+	{
+		if (m_bHasGameStarted) return;
+		m_bHasGameStarted = true;
+		
+		SCR_HintManagerComponent.GetInstance().ShowCustom("Game is starting", "", 10, false);
+
+		if (m_bDisableSpawnAfterStart) 
+		{
+			array<SCR_SpawnPoint> spawnPoints = SCR_SpawnPoint.GetSpawnPoints();
+			foreach(SCR_SpawnPoint sp: spawnPoints)
+			{
+				sp.SetSpawnPointEnabled_S(!sp.IsSpawnPointEnabled());
+			}
+		}
+		OnMatchSituationChanged();
+	}
+	
+	void GameEndCheck() 
+	{
+		if (!m_bHasGameStarted)
+			return;
+		
+		m_iAttackersRemaining = GetFactionRemainingPlayers(m_sAttackerFactionKey);
+		m_iDefendersRemaining = GetFactionRemainingPlayers(m_sDefenderFactionKey);
+		
+		bool attackersDead = m_iAttackersRemaining == 0;
+		bool defendersDead = m_iDefendersRemaining == 0;
+		bool hostageStatus = AllHostagesFreed();
+		
+		PrintFormat("GameEndCheck: A:%1 D:%2 H:%3", attackersDead, defendersDead, hostageStatus, level:LogLevel.SPAM);
+		
+		if (AllHostagesFreed())
+		{
+			Print("All hostages escaped", LogLevel.NORMAL);
+			GameEnd(m_sAttackerFactionKey);
+			return;
+		}
+		
+		if (attackersDead && defendersDead)
+		{
+			GameEndTie();
+			return;
+		}
+		
+		if (attackersDead)
+		{
+			GameEndDefendersWin();
+			return;	
+		}
+		
+		if (defendersDead) 
+		{
+			GameEndAttackersWin();
+			return;
+		}
+	}
+	
+	protected bool AllHostagesFreed()
+	{
+		return GetGame().GetWorld().QueryEntitiesBySphere(
+				GetOrigin(),
+				m_iGameAreaRadius,
+				ProcessFoundHostages,
+				FilterHostageEntities,
+				EQueryEntitiesFlags.DYNAMIC
+		);
+	}
```

Comment:
The query callback/filter/process handlers should be in a separate classes. ACE Backblast has a nice [example](https://github.com/acemod/ACE-Anvil/blob/master/addons/backblast/Scripts/Game/ACE_Backblast/ACE_BackblastComponent.c#L160).